BOLT 12 offer encoding

jkczyz opened the pull request:
WIP on encoding and parsing `offer` messages.
- [ ] Requires upstreaming https://github.com/rust-bitcoin/rust-bech32/pull/66
- [ ] Does not include semantic checks

jkczyz commented on lightning/src/util/ser.rs:
```diff
@@ -510,11 +511,72 @@ impl_array!(3); // for rgb
+/// For variable-length values within TLV record where the length is encoded as part of the record.
+/// Used to prevent encoding the length twice.
+#[derive(Clone)]
+pub(crate) struct WithoutLength<T>(pub T);
```
Ok, so I whipped together a macro that is compliant with the spec. However, if a `vec_type` field is missing when parsing, it will be later serialized as present but with no values (i.e., a parsing-serialization roundtrip results in a different serialization).

jkczyz replied:
> Then we shouldn't write something like that in an `optional_vec` and reject it on the read end?

I guess I'm just not sure if it is technically invalid.

jkczyz commented on lightning/src/offers/mod.rs:
```diff
@@ -0,0 +1,231 @@
+
+use prelude::*;
+
+/// An `offer` TLV stream without any semantic checks, apart from any checks performed when parsing
```
> That's a separate issue though? If we want more errors out of the decoder, we can create those, lets not avoid using our decoder just because we want to add more features to the decoder.

I guess this is related to discussion below about code organization. If `OfferTlvStream` is thought to be an underlying implementation of `Offer`, then there are a few semantic checks related to field presence that would be required:
- `description` is present
- either `node_id` or `paths` is present but not both
- `refund_for` is present only if `send_invoice` is present

I wouldn't want to split these semantic checks across two layers. But if there is only an `Offer` struct, then all these checks would be written in a custom `Readable`, IIUC, where the first is performed by a `read_tlv_fields!` and the other two are handwritten.

> I'm generally very much against this kind of approach - I'm strongly in the "write your datastructures right and the code writes itself" camp - the datastructures should thus generally not be able to represent an invalid state to the greatest extent possible. That implies doing checks during deserialization and refusing to create an object that is invalid, not deserializing and then checking later, leading to tons of unwraps in the accessors assuming the preconditions were checked correctly.

Hmmm... I think there's a tradeoff here. The code won't write itself in this case because we are required to adhere to the encoding as defined in BOLT 12, which doesn't have concepts of enums or structs with optional values, for instance. So the resulting `Readable` and `Writeable` implementations would be very procedural and prone to have inconsistencies.

On the other hand, the `OfferTlvStream` approach leads to a very declarative format, semantic checks are separated from type checks, and serialization is free and therefore less error prone (i.e., doesn't require deconstructing types into component TLV records). Given only one field is required, there should be only one unwrap; the optional field accessors are mostly of the nature: `self.tlv_stream.field.as_ref()`. Additionally, the wrapper `Offer` struct could only be created from a semantically valid `OfferTlvStream` since `Offer`'s decoder would be responsible for performing semantic checks on `OfferTlvStream`. The drawback is accessors for complex types (i.e., enums, structs) may need to access more than one TLV record.

jkczyz replied:
Will push a couple WIP commits showing the accessor API with wrapped types and separated semantic checks. I think I'd prefer to at least have the intermediary struct to enforce the TLV encoding, even if forgoing the wrapper approach. It gives a much better separation between type checks and semantic checsks. The `OfferTlvStream` can always be deconstructed into parts when converting to `Offer`. There's nothing performance critical here.

jkczyz commented on lightning/src/offers/mod.rs:
```diff
@@ -0,0 +1,231 @@
+/// An `offer` TLV stream without any semantic checks, apart from any checks performed when parsing
+/// the underlying types.
+struct OfferTlvStream {
+	_empty: (),
```
Indeed, though note without using `OfferTlvStream` we'd be handwriting `Readable` and `Writeable` with `read_tlv_fields!` and `write_tlv_fields!` instead.

jkczyz commented on lightning/src/offers/mod.rs:
```diff
@@ -0,0 +1,408 @@
+impl Offer {
+	///
+	pub fn description(&self) -> &String {
+		&self.tlv_stream.description.as_ref().unwrap().0
```
> Its exactly this kind of thing that I find really gross. If there's a field that's required, it should be required, not an `Option`. It pushes the complexity into the implementation instead of the struct.

Agreed. Converting the TLV stream instead of wrapping would avoid this and is preferable.

> I'm further not really sure why we can't just make it `required` in the TlvStream - why does it have to be an `option` and then unwraped?

With the converting approach, I would prefer to keep it optional in the TLV stream (but required in `Offer`) so as to not mix error types across layers. Otherwise, missing a description would be a `DecodeError` while missing both `node_id` and `paths` would be a `SemanticError`.

jkczyz replied:
> Right, but there you're trying to work around limitations in the serialization framework by adding complexity further up...instead of improving the serialization framework to do what you want. If there's something you want from the serialization framework, lets add it, not work around it.

Hmmm... I'm not trying to get around any limitations of the serialization framework nor do I want anything more from it. It's working exactly as I expect it should: decoding a stream of TLV records as defined by each record's type. I don't see a compelling reason why decoding should include semantic checks. This is a standard separation of concerns.

Are you arguing that _all_ semantic checks should be handled by the serialization framework? If not, what is the limiting principle as to what should be included there? It doesn't take long before business logic finds it's way in (e.g., `quantity_min <= quantity_max` -- assuming both are set -- or `signature` is valid for the offer given `node_id` or offer has not expired -- assuming `std` ). There are more complex semantics across `amount`, `currency`, and `chains`, which I haven't touched on yet. Excluding all those, supporting even simple semantics would add considerable complexity and maintenance burden to the serialization macros (e.g., if field A is set must not set field B, field A must only be set of field B is set, etc.).

valentinewallace commented on lightning/src/offers/mod.rs:
```diff
@@ -0,0 +1,541 @@
+			),
+		};
+
+		let (paths, node_id) = match self.destination.clone() {
```
@TheBlueMatt may be put off by Vec clones like these so we can do our best for devices that can't take too many memory allocations. In 1503 I found that disallowing extra vec allocations could sometimes change the code materially, so may be worth hashing out which vecs need to go early on.

jkczyz replied:
That was an advantage of (previously) having `Offer` wrap `OfferTlvStream` (no conversion needed), whereas the current approach trades that for having `Offer` always being in a correct state. I think I'm convinced that the non-wrapping version is better. But it would be nice to get rid of this drawback while still having the advantages mentioned in https://github.com/lightningdevkit/rust-lightning/pull/1597/files#r934961472.

Let's consider the use cases:
1. User scans QR code which gets parsed as an `Offer` via `OfferTlvStream` (no duplicate allocs)
2. User constructs an Offer programmatically to serialize as a string for a QR code (duplicate allocs)

The allocs could be removed if instead of using `Display` (which takes `&self`), we have a `From` conversion (which takes `self`) to `OfferTlvStream` or some other user-exposable type like `OfferString` or something. Though they may need to store the `Offer` as well (?), meaning they would need to copy it prior to serialization. Maybe serialization for QR code and in-memory storage would ultimately be different processes? We may need to put some more thought into how this may relate to `offer_id`s and to responding to `invoice_request`s. Let me re-read that part of the spec, but do let me know if you have any thoughts on this already.

jkczyz replied:
Pushed a couple commits at the end for signature verification and (related) `order_id` computation. Another point to consider is that -- while the `order_id` is computed as part of signature verification and can be stored in `Offer` when parsed -- for programmatic constructions, the `order_id` still needs to be computed by converting to the serialized TLV encoding either on-demand or at construction time. So we're going to get hit by allocations there it would seem regardless. Probably could reduce the number of allocations in merkle root computation, though.

jkczyz replied:
> Not sure I'm getting how this could relate to `offer_id`s, is there some way to clarify?

Mostly pointing out when responding to an `invoice_request`, we need to know if we actually know about an `Offer` with the corresponding `offer_id` from the request. So we'll need some way of determining this and responding with an appropriate `invoice`. I guess only tangentially related to allocs.
 
> For `invoice_request`s, I think it could make sense to have something like `Offer.get_invoice_request(amount, quantity, etc) -> Result<InvoiceRequest, Error>`. I was originally thinking this would live in `OnionMessenger` but it probably makes more sense in this module?

Yeah, that was going to be the next step here.

valentinewallace replied:
> Pushed a couple commits at the end for signature verification and (related) `offer_id` computation. Another point to consider is that -- while the `offer_id` is computed as part of signature verification and can be stored in `Offer` when parsed -- for programmatic constructions, the `offer_id` still needs to be computed by converting to the serialized TLV encoding either on-demand or at construction time. So we're going to get hit by allocations there it would seem regardless. Probably could reduce the number of allocations in merkle root computation, though.

Would have to check it out more closely, mostly just wanted to note this as something to consider.

>Mostly pointing out when responding to an invoice_request, we need to know if we actually know about an Offer with the corresponding offer_id from the request. So we'll need some way of determining this and responding with an appropriate invoice. I guess only tangentially related to allocs.

I think we'll need to add something similar to the `inbound_payment` module but for inbound invoices. Then we can do `Offer::verify("payment_secret", &invoice, &inbound_invoice_secret)`. To be clear, the "payment_secret" comes in the `encrypted_data` we encoded for ourselves when we constructed the blinded route, so not 100% on where `verify` lives.

jkczyz commented on lightning/src/offers/mod.rs:
```diff
@@ -0,0 +1,541 @@
+
+///
+#[derive(Clone, Debug)]
+pub struct Offer {
```
Although note that this may require computing the offer id on the fly... hrmm

valentinewallace replied:
Hmm, that's a good point about `offer_id`. Maybe it would be best to make `Offer`s immutable and have a modifiable `OfferConfig` that can be used for easy mutation.

jkczyz replied:
> Ah yeah we might want mutators. I like parsing as `(Offer, Signature)` and having a `Offer.compute_signature()` method, but not an actual struct field. Can discuss in the dev sync. May be useful to start a follow-up issue.

Assuming you have the private key, which wouldn't be the case for a parsed offer. So, FWIW, you'd have to hang on to the signature if you want to re-serialize which is kinda yuck.

valentinewallace replied:
Talked to Matt, I think my vote would be to go with the `lightning-invoice` builder pattern of constructing an `UnsignedOffer` then signing it off at the end to create an immutable `Offer`. Don't feel very strongly though if you have a different idea.

jkczyz replied:
Hmmm... when parsing, how would we know which type to use?

jkczyz replied:
Or are you saying to parse as `(UnsignedOffer, Option<Signature>)`?

valentinewallace replied:
Ah right. I think that points toward having signatures be optional in `Offer`. Talked to Matt, seems it's acceptable for `offer_string.parse<Offer>()` to return a read-only `Offer` with an optional signature. So we'll be giving up the mutability in favor of having an always-valid `Offer`.

jkczyz commented on lightning/src/offers/mod.rs:
```diff
@@ -0,0 +1,579 @@
+#[derive(Clone, Debug)]
+pub struct Offer {
+	id: sha256::Hash,
+	chains: Option<Vec<BlockHash>>,
```
I do recall that but it seems supporting both the main chain and any side chains with the same `offer` is a use case he wants to support as per:  https://github.com/lightning/bolts/pull/798#discussion_r934140378

jkczyz commented on lightning/src/offers/mod.rs:
```diff
@@ -0,0 +1,579 @@
+
+/// An offer parsed from a bech32-encoded string as a TLV stream and the corresponding bytes. The
+/// latter is used for signature verification.
+struct ParsedOffer(OfferTlvStream, Vec<u8>);
```
> I'm really confused here now - so when we're parsing we first read some TLV fields from bytes, then we copy the bytes into this, with the TLV fields moved into it,

The bytes are not copied -- they are moved just as the TLV fields are.

> then we do some analysis of the TLV fields and copy them into a higher-level struct?

I believe they are moved into the higher level struct since `OfferTlvStream` is a value type -- I'd get compiler errors if I tried to reuse any part of the destructured `OfferTlvStream` after it had been moved.

> I'd prefer we just copy the 10 lines of TLV-read code into the top of the `Offer` read method and call it a day, all the copying around of things just seems to add more code and boilerplate and I'm still not sure how it makes anything more readable. Can you help me understand why all the boilerplate here is a simpler design?

I'm not sure how that is materially any different. We'd still need to declare variables prior to those lines, at least some with explicit types. And then repeat the types in the write code, ensuring the vector and integer types match and are serialized correctly rather than relying on the compiler to enforce it.

Additionally, that would require changing the return type to `ParseError` (in order to wrap `DecodeError`) even though all the other errors are `SemanticError`. And then all the error returns would need to be wrapped or littered with `into()`. It would also be possible to return `ParseError` variants which shouldn't be possible at that level.

I pushed a version to show what this looks like. I personally find it much less readable in addition to the other problems that it introduces as mentioned above. Moreover, the `OfferTlvStream` version is much more compact and self-describing, IMHO. I don't see how it introduces boilerplate and more code compared to the alternative.

jkczyz replied:
Discussed this offline yesterday. I pushed a new HEAD commit after reverting the non-`OfferTlvStream` version. The new version avoids any clones when writing. It also hides the `WithoutLength` and `HighZeroBytesDroppedVarInt` types from the `OfferTlvStream` struct definition and consolidates the field and TLV type definitions.

The conversions still need to be aware of the wrapper types, though. Maybe there's a neat way to avoid that with type coercions?

jkczyz replied:
Added some `From` implementations to hide all the `WithoutLength` and `HighZeroBytesDroppedVarInt` wrappers.

jkczyz commented on lightning/src/offers/mod.rs:
```diff
@@ -0,0 +1,579 @@
+	},
+	/// An amount of currency specified using ISO 4712.
+	Currency {
+		/// An ISO 4712 three-letter currency code (e.g., USD).
```
Note that it's already been allocated when the TLV stream was parsed. Extracting characters from strings in Rust without an intermediary `Vec` is kinda gross, too. Either:

```rust
let mut chars = iso4217_code.chars();
let iso4217_code = [chars.next().unwrap(), chars.next().unwrap(), chars.next().unwrap()];
```

Or in 1.55:

```rust
let mut chars = iso4217_code.chars();
[(); ISO4217_CODE_LEN].map(|_| chars.next().unwrap())
```

Happy to change it if you feel strongly. Though note that crates like [iso_4217](https://docs.rs/iso-4217/latest/iso_4217/) work in terms of `str` if we ever need to interface with it.

jkczyz replied:
Yeah, good point. Though it's inside of an enum variant. Not too important at the moment, but changed to `[u8; 3]` anyway.

jkczyz commented on lightning/Cargo.toml:
```diff
@@ -38,7 +38,7 @@ grind_signatures = []
 
 [dependencies]
-bitcoin = { version = "0.28.1", default-features = false, features = ["secp-recovery"] }
+bitcoin = { git = "https://github.com/jkczyz/rust-bitcoin", branch = "2022-06-bech32-version", version = "0.28.1", default-features = false, features = ["secp-recovery"] }
```
Yeah, https://github.com/rust-bitcoin/rust-bech32/pull/66 will need to get in first before a new `rust-bitcoin` release.

jkczyz commented:
@TheBlueMatt @valentinewallace The latest few commits implement `invoice_request` parsing with the `offer` reflected inside. Would be nice to have some early feedback on the approach. Note that each message has a `bytes` field because we need to reflect unknown fields when responding to the message.

jkczyz commented on lightning/src/offers/invoice_request.rs:
```diff
@@ -0,0 +1,111 @@
+
+///
+pub struct InvoiceRequest {
+	bytes: Vec<u8>,
```
When responding to an `invoice_request` with an `invoice`, fields from both the `offer` and the `invoice_request` must be reflected.

jkczyz replied:
We could avoid duplicating the field by always storing the bytes in `Offer` instead of using an empty `Vec` there. I wasn't too thrilled about having it empty depending on which is the outermost message. Likewise, for when parsing an `Invoice`.  And then implement `AsRef<[u8]>` on all messages to pull it from the `Offer`.

jkczyz commented on lightning/src/offers/invoice_request.rs:
```diff
@@ -0,0 +1,111 @@
+	(89, payer_note: String),
+});
+
+impl Bech32Encode for InvoiceRequest {
```
The most recent proposal to the draft is to drop `send_invoice` from `offer` and instead bech32 encode an `invoice_request`. See https://github.com/rustyrussell/lightning-rfc/pull/11.

jkczyz commented:
Ok, so most of the recent changes have been to keep up with https://github.com/rustyrussell/lightning-rfc/pull/11 and https://github.com/rustyrussell/lightning-rfc/pull/7. Probably best to look at all files at once.

In particular, changes include reflecting fields properly when building and `InvoiceRequest` from an `Offer` that may have unknown fields (using the parsed bytes). Would be interested in feedback specifically around that as it may get even more complex when building an `Invoice` from an `InvoiceRequest` with unknown fields since we'd also need to strip off the signature bytes.

jkczyz commented:
Also, note that this doesn't yet have a builder for the "`send_invoice`" `InvoiceRequest` case (i.e., refunds).

jkczyz commented on lightning/src/offers/invoice_request.rs:
```diff
@@ -9,18 +9,121 @@
+		// unknown TLV records, which are not stored in `OfferContents`.
+		let (payer_tlv_stream, _offer_tlv_stream, invoice_request_tlv_stream, _) =
+			self.invoice_request.as_tlv_stream();
+		let offer_bytes = WithoutLength(&self.offer.bytes);
```
Note, this may contain custom records after the signature range. So actually, need to break this into two sequences of bytes. When building an `invoice` from an `invoice_request` will need to three sequences 0-160 (`offer` and `invoice_request`), 240-1000 (signatures to strip), and 1000+ (custom records).

valentinewallace replied:
>  and 1000+ (custom records).

Hm I can't find this in the spec, am I missing it? Commented on the PR: https://github.com/rustyrussell/lightning-rfc/pull/7#discussion_r961244111

jkczyz replied:
Ah, right. As written it does not support it. I suppose unused types in the 0-240 range could be custom, but presumably we may want a dedicated range. BTW, that was suppose to read 0-240 in my earlier comment.

jkczyz commented:
Ok, given the proposed BOLT changes have been merged into https://github.com/lightning/bolts/pull/798, I went ahead and squashed all the commits, rebased, and cleaned things up a bit. There's still a little more to be done on the invoice side, but in the meantime I'm gonna open a new PR with the `Offer` portion (first few commits). I'll leave this PR in draft and update it as I go so the full context is there.

jkczyz commented:
Yeah, everything made it in other PRs. Closing.
